File: packages/lhc-py/src/lhc/shared_tech/storage.py
"""Ported from packages/lhc/src/shared-tech/storage.ts.

`node:sqlite` DatabaseSync → stdlib sqlite3 behind this adapter seam.
WAL-on-open and prepare/get/all/run semantics mirror node:sqlite.
"""

from __future__ import annotations

import sqlite3
from dataclasses import dataclass
from typing import Protocol, runtime_checkable
from weakref import WeakKeyDictionary


@runtime_checkable
class PreparedStatement(Protocol):
    """Mirrors node:sqlite StatementSync prepare/get/all/run semantics."""

    def get(self, *params: object) -> dict[str, object] | None: ...

    def all(self, *params: object) -> list[dict[str, object]]: ...

    def run(self, *params: object) -> object: ...


@runtime_checkable
class Database(Protocol):
    """Adapter seam for node:sqlite DatabaseSync.

    Concrete Phase 2 implementation wraps sqlite3.Connection with WAL on open.
    """

    def prepare(self, sql: str) -> PreparedStatement: ...

    def exec(self, sql: str) -> None: ...

    def close(self) -> None: ...


# Re-export the stdlib type name for call sites that need a concrete annotation
# while the Protocol seam is the public surface.
SqliteConnection = sqlite3.Connection

CURRENT_THREAD_SCHEMA_VERSION = 6

# node:sqlite DatabaseSync path registry (WeakMap).
_database_paths: WeakKeyDictionary[Database, str] = WeakKeyDictionary()


@dataclass(frozen=True, slots=True)
class _StatementRunResult:
    """node:sqlite StatementResult — changes + lastInsertRowid."""

    changes: int
    lastInsertRowid: int


def _split_sql_statements(sql: str) -> list[str]:
    """Split a SQL script into statements without executing a COMMIT.

    Avoids sqlite3.executescript(), which always COMMITs any open transaction
    before running (breaking BEGIN / write / ROLLBACK). Handles quotes and
    line/block comments so semicolons inside them are not separators.
    """
    statements: list[str] = []
    buf: list[str] = []
    i = 0
    n = len(sql)
    in_single = False
    in_double = False
    in_line_comment = False
    in_block_comment = False
    while i < n:
        ch = sql[i]
        nxt = sql[i + 1] if i + 1 < n else ""
        if in_line_comment:
            buf.append(ch)
            if ch == "\n":
                in_line_comment = False
            i += 1
            continue
        if in_block_comment:
            buf.append(ch)
            if ch == "*" and nxt == "/":
                buf.append(nxt)
                i += 2
                in_block_comment = False
                continue
            i += 1
            continue
        if in_single:
            buf.append(ch)
            if ch == "'" and nxt == "'":
                buf.append(nxt)
                i += 2
                continue
            if ch == "'":
                in_single = False
            i += 1
            continue
        if in_double:
            buf.append(ch)
            if ch == '"' and nxt == '"':
                buf.append(nxt)
                i += 2
                continue
            if ch == '"':
                in_double = False
            i += 1
            continue
        if ch == "-" and nxt == "-":
            buf.append(ch)
            buf.append(nxt)
            i += 2
            in_line_comment = True
            continue
        if ch == "/" and nxt == "*":
            buf.append(ch)
            buf.append(nxt)
            i += 2
            in_block_comment = True
            continue
        if ch == "'":
            buf.append(ch)
            in_single = True
            i += 1
            continue
        if ch == '"':
            buf.append(ch)
            in_double = True
            i += 1
            continue
        if ch == ";":
            stmt = "".join(buf).strip()
            if stmt:
                statements.append(stmt)
            buf = []
            i += 1
            continue
        buf.append(ch)
        i += 1
    stmt = "".join(buf).strip()
    if stmt:
        statements.append(stmt)
    return statements


class _SqlitePreparedStatement:
    """Mirrors node:sqlite StatementSync against a sqlite3 connection."""

    def __init__(self, conn: sqlite3.Connection, sql: str) -> None:
        self._conn = conn
        self._sql = sql

    def get(self, *params: object) -> dict[str, object] | None:
        row = self._conn.execute(self._sql, params).fetchone()
        return dict(row) if row is not None else None

    def all(self, *params: object) -> list[dict[str, object]]:
        return [dict(row) for row in self._conn.execute(self._sql, params).fetchall()]

    def run(self, *params: object) -> object:
        # Autocommit when not inside BEGIN (isolation_level=None), matching
        # node:sqlite StatementSync.run — never force-commit mid-transaction.
        cursor = self._conn.execute(self._sql, params)
        changes = cursor.rowcount if cursor.rowcount is not None and cursor.rowcount >= 0 else 0
        last_id = int(cursor.lastrowid) if cursor.lastrowid is not None else 0
        return _StatementRunResult(changes=changes, lastInsertRowid=last_id)


class _SqliteDatabase:
    """Concrete DatabaseSync stand-in: prepare / exec / close + WAL on open."""

    def __init__(self, conn: sqlite3.Connection) -> None:
        self._conn = conn

    def prepare(self, sql: str) -> _SqlitePreparedStatement:
        return _SqlitePreparedStatement(self._conn, sql)

    def exec(self, sql: str) -> None:
        # Must NOT use executescript — it COMMITs any open transaction first,
        # which breaks BEGIN / prepared write / ROLLBACK. Split and execute
        # each statement on the open connection (node:sqlite exec semantics).
        for statement in _split_sql_statements(sql):
            self._conn.execute(statement)

    def close(self) -> None:
        self._conn.close()


def open_database(path: str) -> Database:
    # isolation_level=None → SQLite autocommit (node:sqlite default): each
    # statement commits unless the caller opened a transaction with BEGIN.
    conn = sqlite3.connect(path, isolation_level=None)
    conn.row_factory = sqlite3.Row
    db = _SqliteDatabase(conn)
    _database_paths[db] = path
    db.exec("PRAGMA journal_mode = WAL;")
    db.exec("PRAGMA foreign_keys = ON;")
    db.exec("PRAGMA busy_timeout = 5000;")
    db.exec("PRAGMA synchronous = NORMAL;")
    return db


def database_path_for(db: Database) -> str | None:
    return _database_paths.get(db)


def get_schema_version(db: Database) -> int:
    row = db.prepare("PRAGMA user_version").get()
    if row is None:
        return 0
    value = row.get("user_version", 0)
    return int(value) if value is not None else 0

File: packages/lhc-rs/src/shared_tech/storage.rs
//! Ported from packages/lhc/src/shared-tech/storage.ts. Phase 1 skeleton.
//!
//! REAL adapter surface (Wave 0/1 sqlite seam — fixture exception to the
//! `todo!("phase 2")` skeleton rule):
//! - `open_database` — open + pragma setup
//! - `Db::{path, exec, prepare, close}`
//! - `PreparedStatement::{get, get_params, all, run}`
//! - [`SqlParam`] — crate-owned bind values so tests never name rusqlite
//! - [`StatementRunResult`] — node:sqlite `StatementSync.run` result
//!
//! These mirror the node:sqlite `DatabaseSync` / `StatementSync` surface that
//! `openRaw` and logging-surface fixtures exercise. All other storage helpers
//! remain Phase 2 stubs.
//!
//! BOUNDARY: `rusqlite` may appear only in this file.

use rusqlite::{Connection, ToSql, params_from_iter};

use super::errors::{OpResult, storage_failure};

pub const CURRENT_THREAD_SCHEMA_VERSION: i64 = 13;

/// Crate-owned SQL bind value. Tests and non-storage modules bind through this
/// type — they must not name `rusqlite`.
#[derive(Debug, Clone)]
pub enum SqlParam {
    Null,
    I64(i64),
    F64(f64),
    Text(String),
    Blob(Vec<u8>),
}

impl From<i64> for SqlParam {
    fn from(v: i64) -> Self {
        SqlParam::I64(v)
    }
}

impl From<f64> for SqlParam {
    fn from(v: f64) -> Self {
        SqlParam::F64(v)
    }
}

impl From<String> for SqlParam {
    fn from(v: String) -> Self {
        SqlParam::Text(v)
    }
}

impl From<&str> for SqlParam {
    fn from(v: &str) -> Self {
        SqlParam::Text(v.to_string())
    }
}

impl From<Option<i64>> for SqlParam {
    fn from(v: Option<i64>) -> Self {
        match v {
            Some(n) => SqlParam::I64(n),
            None => SqlParam::Null,
        }
    }
}

impl From<Option<&str>> for SqlParam {
    fn from(v: Option<&str>) -> Self {
        match v {
            Some(s) => SqlParam::Text(s.to_string()),
            None => SqlParam::Null,
        }
    }
}

impl From<Option<String>> for SqlParam {
    fn from(v: Option<String>) -> Self {
        match v {
            Some(s) => SqlParam::Text(s),
            None => SqlParam::Null,
        }
    }
}

impl ToSql for SqlParam {
    fn to_sql(&self) -> rusqlite::Result<rusqlite::types::ToSqlOutput<'_>> {
        match self {
            SqlParam::Null => Ok(rusqlite::types::ToSqlOutput::Owned(
                rusqlite::types::Value::Null,
            )),
            SqlParam::I64(n) => Ok(rusqlite::types::ToSqlOutput::Owned(
                rusqlite::types::Value::Integer(*n),
            )),
            SqlParam::F64(n) => Ok(rusqlite::types::ToSqlOutput::Owned(
                rusqlite::types::Value::Real(*n),
            )),
            SqlParam::Text(s) => Ok(rusqlite::types::ToSqlOutput::Borrowed(
                rusqlite::types::ValueRef::Text(s.as_bytes()),
            )),
            SqlParam::Blob(b) => Ok(rusqlite::types::ToSqlOutput::Borrowed(
                rusqlite::types::ValueRef::Blob(b),
            )),
        }
    }
}

/// node:sqlite `StatementSync.run` result — `{ changes, lastInsertRowid }`.
///
/// Amended Phase 2 Wave 1 repair-r1 (Lee/Fable): the Wave 0 storage seam
/// mirrors `node:sqlite`; this is the documented StatementSync.run result
/// shape, not an invented LHC surface. Callers that need hit/miss semantics
/// read `changes` directly — do **not** use a `SELECT changes()` substitute.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct StatementRunResult {
    pub changes: i64,
    pub last_insert_rowid: i64,
}

/// The one database handle the rest of the crate sees.
///
/// The connection sits behind a `Mutex` so `Db: Sync` — drain/handler futures
/// hold `&Db` across await points and must stay `Send` for `tokio::spawn`
/// (phase-review H5). Uncontended per-statement locking; rusqlite stays
/// confined to this module.
pub struct Db {
    pub(crate) conn: std::sync::Mutex<Connection>,
    path: String,
}

impl Db {
    fn lock(&self) -> std::sync::MutexGuard<'_, Connection> {
        self.conn
            .lock()
            .unwrap_or_else(|poisoned| poisoned.into_inner())
    }
}

/// Minimal node:sqlite `StatementSync` stand-in for fixture probes.
/// REAL methods: [`Self::get`], [`Self::get_params`], [`Self::all`], [`Self::run`].
pub struct PreparedStatement<'a> {
    db: &'a Db,
    sql: String,
}

impl PreparedStatement<'_> {
    fn row_to_map(
        row: &rusqlite::Row<'_>,
    ) -> Result<serde_json::Map<String, serde_json::Value>, String> {
        let names: Vec<String> = row
            .as_ref()
            .column_names()
            .iter()
            .map(|s| (*s).to_string())
            .collect();
        let mut map = serde_json::Map::new();
        for (i, name) in names.iter().enumerate() {
            let value: rusqlite::types::Value = row.get(i).map_err(|err| err.to_string())?;
            map.insert(name.clone(), sqlite_value_to_json(value));
        }
        Ok(map)
    }

    pub fn get(&self) -> Option<serde_json::Map<String, serde_json::Value>> {
        self.get_params(&[])
    }

    /// TS `StatementSync.get(...params)`.
    ///
    /// Error channel (M2 / node:sqlite parity): sqlite prepare/query/step/
    /// row-read failures panic with the underlying sqlite detail (TS throws).
    /// `None` means only "no row", never a swallowed sqlite error.
    pub fn get_params(
        &self,
        params: &[SqlParam],
    ) -> Option<serde_json::Map<String, serde_json::Value>> {
        let conn = self.db.lock();
        let mut stmt = conn
            .prepare(&self.sql)
            .unwrap_or_else(|err| panic!("{err}"));
        let mut rows = stmt
            .query(params_from_iter(params.iter()))
            .unwrap_or_else(|err| panic!("{err}"));
        match rows.next() {
            Ok(Some(row)) => Some(Self::row_to_map(row).unwrap_or_else(|err| panic!("{err}"))),
            Ok(None) => None,
            Err(err) => panic!("{err}"),
        }
    }

    /// TS `StatementSync.all(...params)`.
    ///
    /// Error channel (M2): sqlite failures panic with underlying detail
    /// (TS throws). An empty `Vec` means zero rows, never a swallowed
    /// prepare/query error.
    pub fn all(&self, params: &[SqlParam]) -> Vec<serde_json::Map<String, serde_json::Value>> {
        let conn = self.db.lock();
        let mut stmt = conn
            .prepare(&self.sql)
            .unwrap_or_else(|err| panic!("{err}"));
        let mut rows = stmt
            .query(params_from_iter(params.iter()))
            .unwrap_or_else(|err| panic!("{err}"));
        let mut out = Vec::new();
        loop {
            match rows.next() {
                Ok(Some(row)) => {
                    out.push(Self::row_to_map(row).unwrap_or_else(|err| panic!("{err}")));
                }
                Ok(None) => break,
                Err(err) => panic!("{err}"),
            }
        }
        out
    }

    /// TS `StatementSync.run(...params)` → `{ changes, lastInsertRowid }`.
    ///
    /// Error channel (M2): sqlite failures panic with underlying detail
    /// (TS throws). Returns the exact affected-row count and last insert
    /// row id from this execution (Lee/Fable StatementRunResult amendment).
    pub fn run(&self, params: &[SqlParam]) -> StatementRunResult {
        let conn = self.db.lock();
        let changes = conn
            .execute(&self.sql, params_from_iter(params.iter()))
            .unwrap_or_else(|err| panic!("{err}")) as i64;
        let last_insert_rowid = conn.last_insert_rowid();
        StatementRunResult {
            changes,
            last_insert_rowid,
        }
    }
}

impl Db {
    /// TS `databasePathFor(db)` — path this handle was opened with.
    /// Empty string is a known path (distinct from an absent WeakMap entry).
    pub fn path(&self) -> &str {
        &self.path
    }

    /// TS `DatabaseSync.exec(sql)`. Error channel (M2): sqlite failures panic
    /// with underlying detail (TS throws).
    pub fn exec(&self, sql: &str) {
        self.lock()
            .execute_batch(sql)
            .unwrap_or_else(|err| panic!("{err}"));
    }

    /// TS `DatabaseSync.prepare(sql)`.
    ///
    /// Compiles under the mutex so invalid SQL fails at `prepare` (node:sqlite
    /// parity), then retains the SQL string for later get/all/run execution.
    pub fn prepare(&self, sql: &str) -> PreparedStatement<'_> {
        {
            let conn = self.lock();
            let _stmt = conn.prepare(sql).unwrap_or_else(|err| panic!("{err}"));
        }
        PreparedStatement {
            db: self,
            sql: sql.to_string(),
        }
    }

    /// TS `DatabaseSync.close()`. Close errors panic with underlying detail.
    pub fn close(self) {
        let Db { conn, path: _ } = self;
        let conn = conn
            .into_inner()
            .unwrap_or_else(|poisoned| poisoned.into_inner());
        if let Err((_conn, err)) = conn.close() {
            panic!("{err}");
        }
    }
}

fn sqlite_value_to_json(value: rusqlite::types::Value) -> serde_json::Value {
    match value {
        rusqlite::types::Value::Null => serde_json::Value::Null,
        rusqlite::types::Value::Integer(i) => serde_json::json!(i),
        rusqlite::types::Value::Real(f) => serde_json::json!(f),
        rusqlite::types::Value::Text(s) => serde_json::Value::String(s),
        rusqlite::types::Value::Blob(b) => serde_json::Value::Array(
            b.into_iter()
                .map(|byte| serde_json::Value::Number(byte.into()))
                .collect(),
        ),
    }
}

/// TS `openDatabase(path)`. REAL — Wave 0/1 sqlite seam.
///
/// Error channel (M2): TS `openDatabase` throws on open failure. The frozen
/// Phase 1 public shape wraps open in [`OpResult`] so operation-boundary
/// openers (`ThreadDbOpener`, `HandlerRunContext.open_db`) can return a
/// structured `storage_failure` without panicking. Pragma setup failures
/// still panic (TS would throw after a successful open).
pub fn open_database(path: &str) -> OpResult<Db> {
    match Connection::open(path) {
        Ok(conn) => {
            let db = Db {
                conn: std::sync::Mutex::new(conn),
                path: path.to_string(),
            };
            // busy_timeout FIRST: journal_mode=WAL takes a brief write lock on
            // open, and without a timeout a concurrent opener gets an instant
            // SQLITE_BUSY (observed live: two parallel retrieval tools racing
            // at open — TS 1687d4d / R6).
            db.exec("PRAGMA busy_timeout = 5000;");
            // Only promote to WAL when not already WAL. Re-applying
            // `PRAGMA journal_mode = WAL` on every open takes a write lock that
            // races concurrent openers (TS parity — storage.ts openDatabase).
            // The query form is read-only against an already-WAL file.
            let mode = db
                .prepare("PRAGMA journal_mode")
                .get()
                .and_then(|row| {
                    row.get("journal_mode")
                        .and_then(|v| v.as_str())
                        .map(|s| s.to_ascii_lowercase())
                })
                .unwrap_or_default();
            if mode != "wal" {
                db.exec("PRAGMA journal_mode = WAL;");
            }
            db.exec("PRAGMA foreign_keys = ON;");
            db.exec("PRAGMA synchronous = NORMAL;");
            OpResult::Ok { value: db }
        }
        Err(err) => storage_failure(&err.to_string()),
    }
}

/// Percent-encode a filesystem path for a SQLite URI (keeps `/`, `.`, `_`, `-`).
fn sqlite_uri_encode_path(path: &str) -> String {
    let mut out = String::with_capacity(path.len() + 8);
    for &b in path.as_bytes() {
        match b {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'/' | b'.' | b'_' | b'-' | b'~' => {
                out.push(b as char)
            }
            _ => out.push_str(&format!("%{b:02X}")),
        }
    }
    out
}

/// INTERNAL — side-effect-free open for `peekThreadId`.
///
/// Uses a SQLite URI with `mode=ro` + `immutable=1` so WAL coordination cannot
/// create `-wal`/`-shm` sidecars or write locking state (plain
/// `SQLITE_OPEN_READ_ONLY` is insufficient for WAL files). No pragma/schema
/// mutations. Preserves underlying open-error detail via [`storage_failure`].
/// Not a crate-root or SDK API — `rusqlite` stays here.
///
/// **Contract:** main-file / immutable only — does **not** see uncheckpointed
/// WAL frames. Thread-file identity validation must use
/// [`open_database_for_thread_validation`] instead; do not weaken this opener.
pub(crate) fn open_database_read_only(path: &str) -> OpResult<Db> {
    let uri = format!("file:{}?mode=ro&immutable=1", sqlite_uri_encode_path(path));
    let flags = rusqlite::OpenFlags::SQLITE_OPEN_READ_ONLY
        | rusqlite::OpenFlags::SQLITE_OPEN_URI
        | rusqlite::OpenFlags::SQLITE_OPEN_NO_MUTEX;
    match Connection::open_with_flags(&uri, flags) {
        Ok(conn) => OpResult::Ok {
            value: Db {
                conn: std::sync::Mutex::new(conn),
                path: path.to_string(),
            },
        },
        Err(err) => storage_failure(&err.to_string()),
    }
}

/// INTERNAL — WAL-aware, non-mutating open for `validateThreadFile`.
///
/// Mirrors TS `new DatabaseSync(path, { readOnly: true, timeout })`: a live
/// `mode=ro` URI open **without** `immutable=1`, so uncheckpointed WAL frames
/// are visible and SQLite's own read snapshot keeps identity coherent across a
/// concurrent append or checkpoint. Reads the database in place — no copy, no
/// hash, no temporary directory, no logical byte mutation; only the normal
/// SQLite `-shm` coordination the TypeScript reference already performs.
///
/// Distinct from [`open_database_read_only`], which is the immutable main-only
/// peek seam and cannot see live WAL state — do not reuse it here.
pub(crate) fn open_database_for_thread_validation(path: &str) -> OpResult<Db> {
    // Kept ahead of the open so a missing candidate keeps its exact existing
    // classification independent of sqlite/rusqlite message wording.
    if !std::path::Path::new(path).exists() {
        return storage_failure("unable to open database file");
    }
    let uri = format!("file:{}?mode=ro", sqlite_uri_encode_path(path));
    let flags = rusqlite::OpenFlags::SQLITE_OPEN_READ_ONLY
        | rusqlite::OpenFlags::SQLITE_OPEN_URI
        | rusqlite::OpenFlags::SQLITE_OPEN_NO_MUTEX;
    match Connection::open_with_flags(&uri, flags) {
        Ok(conn) => {
            let db = Db {
                conn: std::sync::Mutex::new(conn),
                path: path.to_string(),
            };
            // Same busy timeout as writers: a read-only probe under WAL still
            // waits briefly when a cross-process writer holds a reserved lock.
            // TS wraps this pragma in try/catch because a read-only handle may
            // refuse some pragmas; the open itself already carries the wait.
            let _ = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
                db.exec("PRAGMA busy_timeout = 5000;");
            }));
            OpResult::Ok { value: db }
        }
        Err(err) => storage_failure(&err.to_string()),
    }
}

/// TS `getSchemaVersion(db)` returns a number (throws on sqlite failure).
/// Frozen Phase 1 shape is [`OpResult<i64>`]: success is `Ok { value }`;
/// sqlite failures on the prepare/get path panic inside [`PreparedStatement`]
/// (TS throw), and are not collapsed into `OpResult::Err`.
pub fn get_schema_version(db: &Db) -> OpResult<i64> {
    let row = db.prepare("PRAGMA user_version").get();
    let version = match row.as_ref().and_then(|m| m.get("user_version")) {
        None | Some(serde_json::Value::Null) => 0,
        Some(serde_json::Value::Number(n)) => n
            .as_i64()
            .or_else(|| n.as_u64().map(|u| u as i64))
            .or_else(|| n.as_f64().map(|f| f as i64))
            .unwrap_or(0),
        Some(other) => other
            .as_i64()
            .or_else(|| other.as_f64().map(|f| f as i64))
            .unwrap_or(0),
    };
    OpResult::Ok { value: version }
}
